File: peer-review/TTK4145-2025-code-peer-review-files/e4490a97/HeisLab2025-Code-snapshot/elev_algo/timer/timer.go
package timer

import (
	"time"
	"github.com//HeisLab2025/elev_algo/elevator_io"
)

var (
	timer_channel = make(chan bool)
	timerEndTime  float64
	timerActive   bool
)

func get_wall_time() float64 {
	return float64(time.Now().UnixNano()) / 1e9
}

func Timer_start(duration float64) {
	go Time(timer_channel)
	timerEndTime = get_wall_time() + duration
	timerActive = true
}

func Timer_stop() {
	timerActive = false
}

func Timer_timed_out() bool {
	return timerActive && get_wall_time() > timerEndTime
}

//sends true on the reciever-channel when door is timed out
func Time(reciever chan<- bool) {
	for {
		if Timer_timed_out() && !elevator_io.GetObstruction() {
			reciever <- true
		}
	}
}

File: peer-review/TTK4145-2025-code-peer-review-files/b67214ff/single_elevator/src/execution/timer.rs
use std::time::{Duration, SystemTime};


pub struct Timer{
    timer_end_time : SystemTime,
    timer_active : bool
}

impl Timer {

pub fn init(timer_end_time : SystemTime, timer_active : bool) -> Self {
    Self {
        timer_end_time,
        timer_active
    }
}

pub fn timer_start(&mut self, duration : Duration) {
    self.timer_end_time = SystemTime::now() + duration;
    self.timer_active = true;
}

pub fn timer_stop(&mut self) {
    self.timer_active = false;
}

pub fn timer_timed_out(&self) -> bool {   
    return self.timer_active && SystemTime::now() > self.timer_end_time;
}

}